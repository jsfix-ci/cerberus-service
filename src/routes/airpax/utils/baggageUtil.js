import * as pluralise from 'pluralise';
import { UNKNOWN_TEXT } from '../../../constants';

const getBaggageWeight = (baggage) => {
  if (!baggage?.weight) {
    return UNKNOWN_TEXT;
  }
  return baggage.weight;
};

const hasBaggage = (targetTask) => {
  return !!targetTask.movement.baggage;
};

const getBaggage = (targetTask) => {
  if (hasBaggage(targetTask)) {
    return targetTask.movement.baggage;
  }
  return null;
};

const getCheckedBags = (baggage, taskDetails = false) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return pluralise.withCount(
    baggage.numberOfCheckedBags,
    '% checked bag',
    '% checked bags',
    (taskDetails ? 'None' : 'No checked bags'),
  );
};

const getNumberOfCheckedBags = (baggage) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return baggage.numberOfCheckedBags;
};

const toFormattedCheckedBagsCount = (baggage) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return pluralise.withCount(
    baggage.numberOfCheckedBags,
    '% bag',
    '% bags',
    'None',
  );
};

const BaggageUtil = {
  get: getBaggage,
  weight: getBaggageWeight,
  checked: getCheckedBags,
  checkedCount: getNumberOfCheckedBags,
  formatCheckedCount: toFormattedCheckedBagsCount,
};

export default BaggageUtil;

export { getBaggageWeight, getBaggage, getCheckedBags, getNumberOfCheckedBags };
