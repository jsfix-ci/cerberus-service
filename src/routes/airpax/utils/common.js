import lookup from 'country-code-lookup';
import { UNKNOWN_TEXT } from '../../../constants';

const convertToIso3Code = (iso2Code) => {
  if (!iso2Code) {
    return UNKNOWN_TEXT;
  }
  if (!lookup.byIso(iso2Code)) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(iso2Code).iso3;
};

const Common = {
  iso3Code: convertToIso3Code,
};

export default Common;

export { convertToIso3Code };
