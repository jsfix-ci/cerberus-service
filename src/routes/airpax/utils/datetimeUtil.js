import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import relativeTime from 'dayjs/plugin/relativeTime';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import { AGO_TEXT, UNKNOWN_TEXT, UTC_DATE_REGEXS } from '../../../constants';

dayjs.extend(utc);
dayjs.extend(relativeTime);
dayjs.extend(customParseFormat);

const validateDate = (date) => {
  return !!(date !== UNKNOWN_TEXT && date && UTC_DATE_REGEXS.some((regex) => regex.test(date)));
};

/**
 * This will return either a formatted UTC datetime or a dayjs datetime object.
 * @param {*} asDayjsObj Flag to determine whether to return a formatted UTC datetime or a dayjs datetime object.
 * @returns A UTC formatted time or a dayjs datetime object.
 */
const getDate = (asDayjsObj = false) => {
  if (asDayjsObj) {
    return dayjs.utc();
  }
  return dayjs.utc().format();
};

const toRelativeTime = (date) => {
  if (!validateDate(date)) {
    return UNKNOWN_TEXT;
  }
  const dateTimeStart = dayjs.utc(date);
  return dateTimeStart.fromNow();
};

const isInPast = (date) => {
  if (!validateDate(date)) {
    return UNKNOWN_TEXT;
  }
  return toRelativeTime(date)?.endsWith(AGO_TEXT);
};

const getFormattedDate = (date, dateFormat) => {
  if (!validateDate(date)) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc(date).format(dateFormat);
};

const formatToUTCDate = (date, inputFormat, outputFormat) => {
  if (!date) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc(date, inputFormat).format(outputFormat);
};

const toDateTimeList = (dateOne, dateTwo) => {
  return [dateOne, dateTwo];
};

const DateTimeUtil = {
  convertToUTC: formatToUTCDate,
  date: getDate,
  format: getFormattedDate,
  isPast: isInPast,
  toList: toDateTimeList,
  relativeTime: toRelativeTime,
  validate: validateDate,
};

export default DateTimeUtil;

export { formatToUTCDate, getDate, getFormattedDate, isInPast, toDateTimeList, toRelativeTime };
