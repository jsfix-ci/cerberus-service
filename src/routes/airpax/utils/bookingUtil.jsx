import React from 'react';
import lookup from 'country-code-lookup';
import { getFormattedDate, toDateTimeList } from './datetimeUtil';
import calculateTimeDifference from '../../../utils/calculateDatetimeDifference';

import {
  UNKNOWN_TEXT,
  SHORT_DATE_FORMAT_ALT,
  STANDARD_HOUR_MINUTE_FORMAT,
  STANDARD_CARD_EXPIRY_FORMAT,
} from '../../../constants';

const getAgentLocation = (agent) => {
  if (!agent?.location) {
    return UNKNOWN_TEXT;
  }
  return agent.location;
};

const getAgentIata = (agent) => {
  if (!agent?.iata) {
    return UNKNOWN_TEXT;
  }
  return agent.iata;
};

const getAgentName = (agent) => {
  if (!agent?.name) {
    return UNKNOWN_TEXT;
  }
  return agent.name;
};

const hasAgent = (booking) => {
  return !!booking?.agent;
};

const getAgent = (booking) => {
  if (hasAgent(booking)) {
    return booking.agent;
  }
  return null;
};

const hasPaymentCard = (payment) => {
  return !!payment?.card;
};

const getPaymentCard = (payment) => {
  if (hasPaymentCard(payment)) {
    return payment.card;
  }
  return null;
};

const getCardExpiry = (payment) => {
  const paymentCard = getPaymentCard(payment);
  if (paymentCard?.expiry) {
    return getFormattedDate(paymentCard.expiry, STANDARD_CARD_EXPIRY_FORMAT);
  }
  return UNKNOWN_TEXT;
};

const getCardLastFourDigits = (payment) => {
  const paymentCard = getPaymentCard(payment);
  if (paymentCard?.number) {
    return paymentCard.number.substr(paymentCard.number.length - 4);
  }
  return UNKNOWN_TEXT;
};

const getPaymentAmount = (payment) => {
  if (!payment?.amount) {
    return UNKNOWN_TEXT;
  }
  return payment.amount;
};

const hasPayments = (booking) => {
  return !!booking?.payments?.length > 0;
};

const getPayments = (booking) => {
  if (hasPayments(booking)) {
    return booking.payments;
  }
  return null;
};

const getTicketType = (ticket) => {
  if (!ticket?.type) {
    return UNKNOWN_TEXT;
  }
  return ticket.type;
};

const getTicketNumber = (ticket) => {
  if (!ticket?.number) {
    return UNKNOWN_TEXT;
  }
  return ticket.number;
};

const hasTicket = (booking) => {
  return !!booking?.ticket;
};

const getTicket = (booking) => {
  if (hasTicket(booking)) {
    return booking.ticket;
  }
  return null;
};

const getBookingType = (booking) => {
  if (!booking?.type) {
    return UNKNOWN_TEXT;
  }
  return booking.type;
};

const getBookingCountryCode = (booking) => {
  if (!booking?.country) {
    return UNKNOWN_TEXT;
  }
  return booking.country;
};

const getBookingCountryName = (booking) => {
  const countryCode = getBookingCountryCode(booking);
  if (countryCode === UNKNOWN_TEXT) {
    return UNKNOWN_TEXT;
  }
  if (lookup.byIso(countryCode) === null) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(countryCode).country;
};

const toBookingDateText = (booking) => {
  if (!booking?.bookedAt) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(booking.bookedAt, SHORT_DATE_FORMAT_ALT);
};

const getBookingReference = (booking) => {
  if (!booking?.reference) {
    return UNKNOWN_TEXT;
  }
  return booking.reference;
};

const getCheckInAt = (booking) => {
  if (!booking?.checkInAt) {
    return UNKNOWN_TEXT;
  }
  return booking.checkInAt;
};

const toCheckInTimeText = (booking) => {
  const checkinPrefix = 'Check-in';
  if (!booking?.checkInAt) {
    return `${checkinPrefix} ${UNKNOWN_TEXT}`;
  }
  return `${checkinPrefix} ${getFormattedDate(booking.checkInAt, STANDARD_HOUR_MINUTE_FORMAT)}`;
};

const getBookedAt = (booking) => {
  return booking?.bookedAt;
};

const getBookedPriorToDeparture = (bookedAt, departureTime) => {
  const dateTimeList = toDateTimeList(bookedAt, departureTime);
  return calculateTimeDifference(dateTimeList);
};

const hasBooking = (targetTask) => {
  return !!targetTask?.movement?.booking;
};

const getBooking = (targetTask) => {
  if (hasBooking(targetTask)) {
    return targetTask.movement.booking;
  }
  return null;
};

const toPaymentsBlock = (booking) => {
  const payments = getPayments(booking);
  if (payments) {
    return payments.map((payment, index) => {
      return (
        <div key={index} className="bottom-border-thin govuk-!-margin-top-1 govuk-!-padding-bottom-1">
          <div className="font__bold">{getPaymentAmount(payment)}</div>
          <div className="font__light">
            Credit card ending {getCardLastFourDigits(payment)}, expiry {getCardExpiry(payment)}
          </div>
        </div>
      );
    });
  }
};

const BookingUtil = {
  get: getBooking,
  bookedAt: getBookedAt,
  checkInAt: getCheckInAt,
  bookedPrior: getBookedPriorToDeparture,
  toCheckInText: toCheckInTimeText,
  bookingRef: getBookingReference,
  toBookingText: toBookingDateText,
  countryCode: getBookingCountryCode,
  countryName: getBookingCountryName,
  bookingType: getBookingType,
  bookingTicket: getTicket,
  ticketNumber: getTicketNumber,
  ticketType: getTicketType,
  payments: getPayments,
  containsPayments: hasPayments,
  paymentAmount: getPaymentAmount,
  paymentCard: getPaymentCard,
  cardLastFourDigits: getCardLastFourDigits,
  cardExpiry: getCardExpiry,
  agent: getAgent,
  agentName: getAgentName,
  agentIata: getAgentIata,
  agentLocation: getAgentLocation,
  paymentsBlock: toPaymentsBlock,
};

export default BookingUtil;

export {
  toBookingDateText,
  getBookingReference,
  toCheckInTimeText,
  getBooking,
  getBookedAt,
  getCheckInAt,
  getBookedPriorToDeparture,
  getBookingCountryCode,
  getBookingCountryName,
  getBookingType,
  getTicket,
  getTicketNumber,
  getTicketType,
  getPayments,
  getPaymentAmount,
  getPaymentCard,
  getCardLastFourDigits,
  getCardExpiry,
  getAgent,
  getAgentName,
  getAgentIata,
  getAgentLocation,
  toPaymentsBlock,
};
