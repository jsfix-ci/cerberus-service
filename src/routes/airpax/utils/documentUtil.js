import moment from 'moment';
import lookup from 'country-code-lookup';
import {
  A_TITLE_CASE_TEXT,
  AN_TITLE_CASE_TEXT,
  AGO_TEXT,
  A_SMALL_TEXT,
  AN_SMALL_TEXT,
  UNKNOWN_TEXT,
  AFTER_TRAVEL_TEXT,
  BEFORE_TRAVEL_TEXT,
  SHORT_DATE_FORMAT_ALT,
} from '../../../constants';
import { formatField } from '../../../utils/formatField';
import { getFormattedDate } from './datetimeUtil';

const calculateExpiry = (passportExpiry, arrivalTime) => {
  const expiry = arrivalTime
    && passportExpiry !== 'Unknown'
    && `${arrivalTime},${moment(passportExpiry).format('YYYY-MM-DDTHH:mm:ss')}`;
  if (expiry) {
    return formatField('BOOKING_DATETIME', expiry)
      .split(', ')[1]
      .replace(BEFORE_TRAVEL_TEXT, AFTER_TRAVEL_TEXT)
      .replace(AGO_TEXT, BEFORE_TRAVEL_TEXT)
      .replace(A_SMALL_TEXT, A_TITLE_CASE_TEXT)
      .replace(AN_SMALL_TEXT, AN_TITLE_CASE_TEXT);
  }
  return UNKNOWN_TEXT;
};

const hasDocument = (person) => {
  return !!person?.document;
};

const getDocument = (person) => {
  if (hasDocument(person)) {
    return person.document;
  }
  return null;
};

const getDocumentExpiry = (document, taskDetails = false) => {
  const expiryPrefix = 'Expires';
  if (!document?.expiry) {
    return taskDetails ? UNKNOWN_TEXT : `${expiryPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}` : `${expiryPrefix} ${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentValidity = (document, taskDetails = false) => {
  const validityPrefix = 'Valid from';
  if (!document?.validFrom) {
    return taskDetails ? UNKNOWN_TEXT : `${validityPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}` : `${validityPrefix} ${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentType = (document) => {
  if (!document?.type) {
    return UNKNOWN_TEXT;
  }
  return document.type;
};

const getDocumentNumber = (document) => {
  if (!document?.number) {
    return UNKNOWN_TEXT;
  }
  return document.number;
};

const getDocumentName = (person) => {
  if (!person?.name) {
    return UNKNOWN_TEXT;
  }
  const firstNames = person?.name?.first || UNKNOWN_TEXT;
  const lastName = person?.name?.last || UNKNOWN_TEXT;
  return `${lastName.toUpperCase()}, ${firstNames}`;
};

// TODO finish implementation once data flows through
const getDocumentIdentification = (document) => {
  if (!document) {
    return UNKNOWN_TEXT;
  }
  return UNKNOWN_TEXT;
};

const getDocumentCountryOfIssue = (document, taskDetails = false) => {
  const countryOfIssuePrefix = 'Issued by';
  if (!document?.countryOfIssue) {
    return taskDetails ? UNKNOWN_TEXT : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }
  if (taskDetails) {
    return lookup.byIso(document.countryOfIssue) !== null
      ? `${lookup.byIso(document.countryOfIssue).country} (${document.countryOfIssue})`
      : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }

  return lookup.byIso(document.countryOfIssue) !== null
    ? `${countryOfIssuePrefix} ${document.countryOfIssue}`
    : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
};

const getDocumentCountryOfIssueCode = (document) => {
  if (!document?.countryOfIssue) {
    return UNKNOWN_TEXT;
  }
  return document.countryOfIssue;
};

const getDocumentNationality = (document, taskDetails = false) => {
  return getDocumentCountryOfIssue(document, taskDetails);
};

const getDocumentDOB = (document) => {
  if (!document?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(document?.dateOfBirth, SHORT_DATE_FORMAT_ALT);
};

const DocumentUtil = {
  get: getDocument,
  docExpiry: getDocumentExpiry,
  docValidity: getDocumentValidity,
  docType: getDocumentType,
  docNumber: getDocumentNumber,
  docName: getDocumentName,
  docIdentification: getDocumentIdentification,
  docCountry: getDocumentCountryOfIssue,
  docCountryCode: getDocumentCountryOfIssueCode,
  docNationality: getDocumentNationality,
  docDOB: getDocumentDOB,
  calculateExpiry,
};

export default DocumentUtil;

export {
  getDocument,
  getDocumentExpiry,
  getDocumentValidity,
  getDocumentType,
  getDocumentNumber,
  getDocumentName,
  getDocumentIdentification,
  getDocumentCountryOfIssue,
  getDocumentCountryOfIssueCode,
  getDocumentNationality,
  getDocumentDOB,
  calculateExpiry,
};
