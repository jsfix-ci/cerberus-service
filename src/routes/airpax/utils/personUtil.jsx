import React from 'react';
import lookup from 'country-code-lookup';
import dayjs from 'dayjs';
import formatGender from '../../../utils/genderFormatter';
import { getFormattedDate } from './datetimeUtil';
import { SHORT_DATE_FORMAT_ALT, UNKNOWN_TEXT } from '../../../constants';

const getNationality = (person) => {
  if (!person?.nationality) {
    return UNKNOWN_TEXT;
  }
  return person.nationality;
};

const getCountryName = (person) => {
  if (!person?.nationality) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(person.nationality).country;
};

const getDateOfBirth = (person) => {
  if (!person?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(person.dateOfBirth, SHORT_DATE_FORMAT_ALT);
};

const getAge = (person) => {
  if (!person?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc().diff(dayjs(person.dateOfBirth), 'year');
};

const getGender = (person) => {
  if (!person) {
    return UNKNOWN_TEXT;
  }
  return formatGender(person.gender);
};

const getLastName = (person, capitalize = true) => {
  if (!person?.name?.last) {
    return capitalize ? UNKNOWN_TEXT.toUpperCase() : UNKNOWN_TEXT;
  }
  return capitalize ? person.name.last.toUpperCase() : person.name.last;
};

const getFirstName = (person) => {
  if (!person?.name?.first) {
    return UNKNOWN_TEXT;
  }
  return person.name.first;
};

const getFrequentFlyerNumber = (person) => {
  if (!person?.frequentFlyerNumber) {
    return UNKNOWN_TEXT;
  }
  return person.frequentFlyerNumber;
};

const hasSSRCodes = (person) => {
  return !!person?.ssrCodes?.length > 0;
};

const getSSRCodes = (person) => {
  if (!hasSSRCodes(person)) {
    return UNKNOWN_TEXT;
  }
  return person.ssrCodes.join(', ');
};

const toCoTravellers = (otherPersons) => {
  if (!otherPersons) {
    return <li className="govuk-!-font-weight-bold">None</li>;
  }
  const maxToDisplay = 4;
  const remaining = otherPersons.length > maxToDisplay ? otherPersons.length - maxToDisplay : 0;
  return otherPersons.map((person, index) => {
    if (index < maxToDisplay) {
      return (
        <li key={index} className="govuk-!-font-weight-bold">
          {getLastName(person)}, {getFirstName(person)}{(index !== maxToDisplay - 1)
                && (index !== otherPersons.length - 1) ? ',' : ''} {(remaining > 0 && index + 1 === maxToDisplay)
                  ? ` plus ${remaining} more` : ''}
        </li>
      );
    }
  });
};

const getAllPersons = (person, otherPersons) => {
  let allPersons = [];
  if (person) {
    allPersons.push(person);
  }
  if (otherPersons) {
    allPersons = allPersons.concat(otherPersons);
  }
  return allPersons;
};

const getTotalNumberOfPersons = (targetTask) => {
  if (!targetTask?.movement?.person) {
    return 0;
  }
  return targetTask.movement.otherPersons.length + 1;
};

const hasOtherPersons = (targetTask) => {
  return !!targetTask?.movement?.otherPersons?.length;
};

const getOtherPersons = (targetTask) => {
  if (hasOtherPersons(targetTask)) {
    return targetTask.movement.otherPersons;
  }
  return null;
};

const hasPerson = (targetTask) => {
  return !!targetTask?.movement?.person;
};

const getPerson = (targetTask) => {
  if (hasPerson(targetTask)) {
    return targetTask.movement.person;
  }
  return null;
};

const PersonUtil = {
  get: getPerson,
  getOthers: getOtherPersons,
  totalPersons: getTotalNumberOfPersons,
  toOthers: toCoTravellers,
  allPersons: getAllPersons,
  firstname: getFirstName,
  lastname: getLastName,
  gender: getGender,
  dob: getDateOfBirth,
  age: getAge,
  nationality: getNationality,
  countryName: getCountryName,
  frequentFlyerNumber: getFrequentFlyerNumber,
  ssrCodes: getSSRCodes,
};

export default PersonUtil;

export {
  getNationality,
  getCountryName,
  getDateOfBirth,
  getAge,
  getPerson,
  getGender,
  getFirstName,
  getLastName,
  toCoTravellers,
  getTotalNumberOfPersons,
  getOtherPersons,
  getFrequentFlyerNumber,
  getSSRCodes,
};
