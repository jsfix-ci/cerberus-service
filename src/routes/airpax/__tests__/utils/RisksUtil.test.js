import renderer from 'react-test-renderer';
import { RisksUtil } from '../../utils';

describe('RisksUtil', () => {
  it('should get risk if present', () => {
    const targetListDataMin = {
      id: 'DEV-20220415-001',
      risks: {
        targetingIndicators: {
          indicators: [
            {
              description: 'Quick turnaround freight (under 24 hours)',
            },
            {
              description: 'Quick turnaround tourist (under 24 hours)',
            },
          ],
          count: 2,
          score: 60,
        },
      },
    };

    const expected = {
      targetingIndicators: {
        indicators: [
          {
            description: 'Quick turnaround freight (under 24 hours)',
          },
          {
            description: 'Quick turnaround tourist (under 24 hours)',
          },
        ],
        count: 2,
        score: 60,
      },
    };

    const output = RisksUtil.getRisks(targetListDataMin);
    expect(output).toEqual(expected);
  });

  it('should get targeting indicators if present', () => {
    const risks = {
      targetingIndicators: {
        indicators: [
          {
            description: 'Quick turnaround freight (under 24 hours)',
          },
          {
            description: 'Quick turnaround tourist (under 24 hours)',
          },
        ],
        count: 2,
        score: 60,
      },
    };

    const expected = {
      indicators: [
        {
          description: 'Quick turnaround freight (under 24 hours)',
        },
        {
          description: 'Quick turnaround tourist (under 24 hours)',
        },
      ],
      count: 2,
      score: 60,
    };

    const output = RisksUtil.getIndicators(risks);
    expect(output).toEqual(expected);
  });

  it('should return a formatted targeting indicators if present', () => {
    const targetingIndicators = {
      indicators: [
        {
          description: 'Quick turnaround freight (under 24 hours)',
        },
        {
          description: 'Quick turnaround tourist (under 24 hours)',
        },
      ],
      count: 2,
      score: 60,
    };

    const tree = renderer.create(RisksUtil.format(targetingIndicators)).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render No warnings text if warning status is NO', () => {
    const selector = {
      id: 279,
      reference: '2021-279 A',
      category: 'A',
      warning: {
        status: 'NO',
        types: [

        ],
        detail: null,
      },
    };

    const output = RisksUtil.getWarning(selector);
    expect(output).toEqual('No warnings');
  });

  it('should render Warnings currently unavailable text if warning status is Currently unavailable', () => {
    const selector = {
      id: 279,
      reference: '2021-279 A',
      category: 'A',
      warning: {
        status: 'Currently unavailable',
        types: [

        ],
        detail: null,
      },
    };

    const output = RisksUtil.getWarning(selector);
    expect(output).toEqual('Warnings currently unavailable');
  });

  it('should render comma seperated warnings if warning status is YES', () => {
    const selector = {
      id: 279,
      reference: '2021-279 A',
      category: 'A',
      warning: {
        status: 'YES',
        types: ['VIOLENCE', 'FIREARMS', 'WEAPONS', 'CONTAGION', 'SELF_HARM'],
        detail: null,
      },
    };

    const output = RisksUtil.getWarning(selector);
    expect(output).toEqual('Violence, Firearms, Weapons, Contagion, Self harm');
  });

  it('should render nothing if warning status is null', () => {
    const selector = {
      id: 279,
      reference: '2021-279 A',
      category: 'A',
      warning: {
        status: null,
        types: [],
        detail: null,
      },
    };

    const output = RisksUtil.getWarning(selector);
    expect(output).toEqual(undefined);
  });

  it('should get indicator matches if present', () => {
    const selector = {
      indicatorMatches: [
        {
          entity: 'Message',
          descriptor: 'mode',
          operator: 'in',
          value: 'RORO Accompanied Freight',
        },
        {
          entity: 'Trailer',
          descriptor: 'registrationNumber',
          operator: 'equal',
          value: 'qwerty',
        },
      ],
    };

    const expected = [
      {
        descriptor: 'mode',
        entity: 'Message',
        operator: 'in',
        value: 'RORO Accompanied Freight',
      },
      {
        descriptor: 'registrationNumber',
        entity: 'Trailer',
        operator: 'equal',
        value: 'qwerty',
      },
    ];
    const output = RisksUtil.getMatches(selector);
    expect(output).toEqual(expected);
  });

  it('should return empty array if indicator matches does not contain any indicators', () => {
    const selector = {
      indicatorMatches: [],
    };

    const output = RisksUtil.getMatches(selector);
    expect(output).toEqual([]);
  });

  it('should return null if indicator matches is null', () => {
    const selector = {
      indicatorMatches: null,
    };

    const output = RisksUtil.getMatches(selector);
    expect(output).toEqual(null);
  });

  it('should get rules matches if present', () => {
    const rules = {
      risks: {
        matchedRules: [
          {
            id: 535,
            name: 'Selector Matched Rule',
            type: 'Both',
            priority: 'Tier 1',
            agency: '',
            description: 'Test Description',
            version: 1,
            abuseTypes: [
              'National Security at the Border',
            ],
            indicatorMatches: [
              {
                entity: 'Message',
                descriptor: 'mode',
                operator: 'in',
                value: '[RORO Accompanied Freight, RORO Tourist, RORO Unaccompanied Freight]',
              },
              {
                entity: 'Message',
                descriptor: 'selectorsMatched',
                operator: 'equal',
                value: 'true',
              },
            ],
          },
        ],
      },
    };

    const expected = [
      { 'abuseTypes': ['National Security at the Border'],
        'agency': '',
        'description': 'Test Description',
        'id': 535,
        'indicatorMatches': [{ 'descriptor': 'mode', 'entity': 'Message', 'operator': 'in', 'value': '[RORO Accompanied Freight, RORO Tourist, RORO Unaccompanied Freight]' }, { 'descriptor': 'selectorsMatched', 'entity': 'Message', 'operator': 'equal', 'value': 'true' }],
        'name': 'Selector Matched Rule',
        'priority': 'Tier 1',
        'type': 'Both',
        'version': 1 }];

    const output = RisksUtil.getRules(rules);
    expect(output).toEqual(expected);
  });

  it('should return null if rules matches is null', () => {
    const rules = {
      risks: {
        matchedRules: null,
      },
    };

    const output = RisksUtil.getRules(rules);
    expect(output).toEqual(null);
  });

  it('should get target task highestThreatLevel if present', () => {
    const risks = {
      highestThreatLevel: {
        type: 'SELECTOR',
        value: 'B',
      },
    };

    const output = RisksUtil.getHighestThreat(risks);
    expect(output).toEqual(risks.highestThreatLevel);
  });

  it('should return null if target task highestThreatLevel is null', () => {
    const risks = {
      highestThreatLevel: null,
    };

    const output = RisksUtil.getHighestThreat(risks);
    expect(output).toEqual(risks.highestThreatLevel);
  });

  it('should get target task matchedSelectorGroups if present', () => {
    const risks = {
      matchedSelectorGroups: {
        groups: [
          {
            groupReference: 'SR-245',
            groupVersionNumber: 1,
            category: 'A',
            threatType: 'Class A Drugs',
            selectors: [
              {
                id: 279,
                reference: '2021-279',
                category: 'A',
                warning: {
                  status: 'YES',
                  types: [
                    'VIOLENCE',
                  ],
                  detail: 'other warning details',
                },
                indicatorMatches: [],
                description: 'RORO Accompanied Freight qwerty',
              },
            ],
          },
        ],
        totalNumberOfSelectors: 1,
      },
    };

    const output = RisksUtil.getMatchedSelectorGroups(risks);
    expect(output).toEqual(risks.matchedSelectorGroups);
  });

  it('should return null if target task matchedSelectorGroups is null', () => {
    const risks = {
      matchedSelectorGroups: null,
    };

    const output = RisksUtil.getMatchedSelectorGroups(risks);
    expect(output).toEqual(risks.matchedSelectorGroups);
  });

  it('should get target task matchedRules if present', () => {
    const risks = {
      matchedRules: [
        {
          id: 535,
          name: 'Selector Matched Rule',
          type: 'Both',
          priority: 'Tier 1',
          version: 1,
          abuseTypes: [
            'National Security at the Border',
          ],
          indicatorMatches: [],
        },
      ],
    };

    const output = RisksUtil.getMatchedRules(risks);
    expect(output).toEqual(risks.matchedRules);
  });

  it('should return null if target task matchedRules is null', () => {
    const risks = {
      matchedRules: null,
    };

    const output = RisksUtil.getMatchedRules(risks);
    expect(output).toEqual(risks.matchedRules);
  });
});
