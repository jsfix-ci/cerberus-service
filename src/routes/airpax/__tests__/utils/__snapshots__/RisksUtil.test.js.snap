// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RisksUtil should return a formatted targeting indicators if present 1`] = `
<ul
  className="govuk-list item-list--bulleted"
>
  <li
    className="govuk-!-font-weight-bold govuk-!-font-size-16"
  >
    2 indicators
  </li>
  <li
    className="threat-indicator-bullet govuk-!-font-size-16"
  >
    Quick turnaround freight (under 24 hours)
  </li>
  <li
    className="threat-indicator-bullet govuk-!-font-size-16"
  >
    Quick turnaround tourist (under 24 hours)
  </li>
</ul>
`;
